/**
 * GossipSampling
 * Copyright (C) Matthew Love 2024 (gossipsampling@gmail.com)
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * 
 */

syntax = "proto3";

import "google/protobuf/empty.proto";

package gossip;

service GossipProtocol {
    rpc PushView(ViewProto) returns (google.protobuf.Empty) {}
    rpc PullView(google.protobuf.Empty) returns (ViewProto) {}
    rpc PushPullView(ViewProto) returns (ViewProto) {}
}


/* Expand as Needed, proto3 allows us to make every field optional */
message NodeDescriptorProto {
    string address = 1;
    uint32 age = 2; /* Base Gossip Protocol */
}

message ViewProto {
    repeated NodeDescriptorProto nodes = 1;
}